airbyte-integrations/connectors/source-github/source_github/streams.py

```diff
@@ -75,11 +89,34 @@ def read_records(self, **kwargs) -> Iterable[Mapping[str, Any]]:
             # function we have `response.raise_for_status()` so we don't have much choice on how to handle errors.
-            # We added this try/except code because for private repositories `Teams` stream is not available and we get
-            # "404 Client Error: Not Found for url: https://api.github.com/orgs/sherifnada/teams?per_page=100" error.
-            # Blocked on https://github.com/airbytehq/airbyte/issues/3514.
-            if "/teams?" in error_msg:
-                error_msg = f"Syncing Team stream isn't available for repository {self.repository}"
+            # Bocked on https://github.com/airbytehq/airbyte/issues/3514.
+            if "403 Client Error" in error_msg:
+                error_msg = (
+                    f"Syncing `{self.__class__.__name__}` stream isn't available for repository "
+                    f"`{self.repository}` and your `access_token`, seems like you don't have permissions for "
+                    f"this stream."
+                )
+            elif "404 Client Error" in error_msg and "/teams?" in error_msg:
+                # For private repositories `Teams` stream is not available and we get "404 Client Error: Not Found for
+                # url: https://api.github.com/orgs/sherifnada/teams?per_page=100" error.
+                error_msg = f"Syncing `Team` stream isn't available for repository `{self.repository}`."
+            elif "502 Server Error" in error_msg:
+                # Sometimes we may receive "502 Server Error: Bad Gateway for url:" errors. Normal behaviour in to just
+                # repeat request later, but not with GitHub API because we may create an infinite loop. For large
+                # streams (like `comments`) we are providing following parameters: `since`, `sort` and `direction`.
+                # 502 error can occur when you specify very distant start_date in the past. For example, if you try to
+                # do request to the following url:
+                # https://api.github.com/repos/apache/superset/issues/comments?per_page=100&page=1&since=2015-01-01T00:00:00Z&sort=updated&direction=desc
+                # you will be getting `502 Server Error` in 95% of responses. This is because `since` value is very
+                # distant in the past and apparently GitHub can't handle it. So since we handle streams synchronously
+                # we may stuck in this loop for a long time (hours). That's why it's better to skip stream and show
+                # message to user explaining what went wrong. Also explanation will be added to documentation for
+                # GitHub connector.
+                error_msg = (
+                    f"Syncing `{self.__class__.__name__}` stream isn't available right now. We got 502 error "
+                    f"code from GitHub API meaning that GitHub has some issues while processing request. "
+                    f"You may try again later or specify more recent `start_date` parameter."
+                )
```

Comment:
Well, this error catching seems really odd, is it the only way to catch them? I mean like checking it by status code or smth?

Reply:
Yes, we could check by status code, but it will still be the same code with `if` statements like:

```
if e.response.status_code == 403:
    ...
elif e.response.status_code == 404 and "/teams?" in error_msg:
    ....
elif e.response.status_code == 502:
    ....
```

Reply:
Done.